scripts/globals/job_utils/beastmaster.lua

```diff
@@ -0,0 +1,132 @@
+
+local function getCharmDuration(charmer, target)
+    local charmDuration = 0
+
+    -- Calculate base duration and dLvl
+    local baseCharmDuration = math.floor(1.25 * charmer:getStat(xi.mod.CHR) + 150)
+    local dLvl = charmer:getMainLvl() - target:getMainLvl()
+
+    -- Default multiplier for dLvl -6 or lower
+    local dLvlCharmMult = 1 / 24
+
+    if dLvl >= -6 and dLvl < 9 then
+        -- Quintic fitting of duration multiplier as function of dLvl (r^2 > 0.999)
+        dLvlCharmMult = 0.9997336 + 0.3652882 * dLvl + 0.02097742 * dLvl^2
+                    - 0.004106429 * dLvl^3 + 0.000007231037 * dLvl^4
+                    + 0.00005102634 * dLvl^5
+    -- Caps at dLvl > 9
+    elseif dLvl >= 9 then
+        dLvlCharmMult = 6
+    end
+
+    -- Apply the dLvl multiplier
+    charmDuration = baseCharmDuration * dLvlCharmMult
+
+    -- Apply charm duration extension from gear
+    local charmTimeMod = charmer:getMod(xi.mod.CHARM_TIME)
+    local extraDurationFromMod = charmDuration * (charmTimeMod * 0.5 / 10) -- Assumes 5% per charmTimeMod
+    charmDuration = charmDuration + extraDurationFromMod
+
+    return math.floor(charmDuration)
+end
+
+xi.job_utils.beastmaster.getCharmChance = function(charmer, target, includeMods)
+    if
+        not charmer or                                -- Invalid charmer
+        not target or                                 -- Invalid target
+        not charmer:isPC() or                         -- Charmer not a player
+        not target:isMob() or                         -- Target not a mob
+        target:getMobMod(xi.mobMod.CHARMABLE) == 0 or -- Not charmable
+        (target:getMaster() ~= nil)                   -- Someone else's pet
+    then
+        return 0
+    end
+
+    -- Use the players BST level (even if subjob) for charm chance calc
+    local charmerJobLevel = charmer:getJobLevel(xi.job.BST)
+    local targetLevel     = target:getMainLvl()
+    local charmres        = target:getMod(xi.mod.CHARMRES)
+    local charmChance     = 50 - charmres
+    -- dLvl only applies when player lvl < mob lvl
+    -- and varies for different target levels
+    if charmerJobLevel < targetLevel
+        if targetLevel >= 71 then
+            charmChance = charmChance - 10 * (targetLevel - charmerJobLevel)
+        elseif targetLevel >= 51 then
+            charmChance = charmChance - 5 * (targetLevel - charmerJobLevel)
+        else
+            charmChance = charmChance - 3 * (targetLevel - charmerJobLevel)
+        end
+    end
+
+    -- Another multiplier determined by target light res rank
+    -- as charm is a light based ability
+    local rank = target:getMod(xi.mod.LIGHT_RES_RANK)
+    if rank <= -3 then
+        charmChance = charmChance * 1.5
+    elseif rank <= -2 then
+        charmChance = charmChance * 1.4
+    elseif rank <= -1 then
+        charmChance = charmChance * 1.2
+    elseif rank <= 0 then
+        charmChance = charmChance
+    else
+        charmChance = charmChance / 2
+    end
```

Comment:
This seems weird to me.
If resistance rank is...
- 10, shouldnt MHR be floored to 5%?
- 11, shouldnt it auto-fail?

All of this looks like it COULD use the same functions used in magic enfeebles, resistance wise.

Reply:
> This seems weird to me. If resistance rank is...
> 
> * 10, shouldnt MHR be floored to 5%?
> * 11, shouldnt it auto-fail?
> 
> All of this looks like it COULD use the same functions used in magic enfeebles, resistance wise.

I think the highest resist rank validated on retail was a mob with rank of 2 (this [mob](https://www.bg-wiki.com/ffxi/Category:Twitherym) via [this post](http://blog.livedoor.jp/yotsuba_yudaya/archives/2007204.html#more)), it might be possible to check a mob with rank of 4 (Sabotender) but I think most mobs with higher ranks than that are not susceptible to charm for testing.

Reply:
> > This seems weird to me. If resistance rank is...
> > 
> > * 10, shouldnt MHR be floored to 5%?
> > * 11, shouldnt it auto-fail?
> > 
> > All of this looks like it COULD use the same functions used in magic enfeebles, resistance wise.
> 
> I think the highest resist rank validated on retail was a mob with rank of 2 (this [mob](https://www.bg-wiki.com/ffxi/Category:Twitherym) via [this post](http://blog.livedoor.jp/yotsuba_yudaya/archives/2007204.html#more)), it might be possible to check a mob with rank of 4 (Sabotender) but I think most mobs with higher ranks than that are not susceptible to charm for testing. Maybe there is another way of testing or creating high rank mobs?

After talking with Jimmayus, if there is some kind of flooring for charm it is not possible to determine (on retail) because from our understanding there are not charmable mobs with light EEM lower than 50%. So makes sense to stick with the PR values as those have been validated in captures.

Reply:
> > This seems weird to me. If resistance rank is...
> > 
> > * 10, shouldnt MHR be floored to 5%?
> > * 11, shouldnt it auto-fail?
> > 
> > All of this looks like it COULD use the same functions used in magic enfeebles, resistance wise.
> 
> I think the highest resist rank validated on retail was a mob with rank of 2 (this [mob](https://www.bg-wiki.com/ffxi/Category:Twitherym) via [this post](http://blog.livedoor.jp/yotsuba_yudaya/archives/2007204.html#more)), it might be possible to check a mob with rank of 4 (Sabotender) but I think most mobs with higher ranks than that are not susceptible to charm for testing. Maybe there is another way of testing or creating high rank mobs?

After talking with Jimmayus, if there is some kind of flooring for charm it is not possible to determine (on retail) because from our understanding there are not charmable mobs with light EEM lower than 50%. So makes sense to stick with the PR values as those have been validated in captures.